firedrake/assemble.py

```diff
@@ -91,21 +92,114 @@ def assemble(expr, tensor=None, bcs=None, form_compiler_parameters=None,
         raise TypeError("Unknown keyword arguments '%s'" % ', '.join(kwargs.keys()))
 
     if isinstance(expr, (ufl.form.Form, slate.TensorBase)):
-        loops = _assemble(expr, tensor=tensor, bcs=solving._extract_bcs(bcs),
-                          form_compiler_parameters=form_compiler_parameters,
-                          mat_type=mat_type,
-                          sub_mat_type=sub_mat_type, appctx=appctx,
-                          diagonal=diagonal,
-                          options_prefix=options_prefix)
-        for l in loops:
-            m = l()
-        return m
+        mat_type, sub_mat_type = get_mat_type(mat_type, sub_mat_type, expr.arguments())
+        assembly_rank = _get_assembly_rank(expr, diagonal)
+        if assembly_rank == AssemblyRank.SCALAR:
+            return _assemble_scalar(expr, tensor, bcs=bcs, form_compiler_parameters=form_compiler_parameters, mat_type=mat_type, sub_mat_type=sub_mat_type, appctx=appctx, options_prefix=options_prefix, assemble_now=assemble_now)
+        elif assembly_rank == AssemblyRank.VECTOR:
+            return _assemble_vector(expr, tensor, bcs=bcs, form_compiler_parameters=form_compiler_parameters, mat_type=mat_type, sub_mat_type=sub_mat_type, diagonal=diagonal, appctx=appctx, options_prefix=options_prefix, assemble_now=assemble_now)
+        elif assembly_rank == AssemblyRank.MATRIX:
+            return _assemble_matrix(expr, tensor, bcs=bcs, form_compiler_parameters=form_compiler_parameters, mat_type=mat_type, sub_mat_type=sub_mat_type, appctx=appctx, options_prefix=options_prefix, assemble_now=assemble_now)
     elif isinstance(expr, ufl.core.expr.Expr):
         return assemble_expressions.assemble_expression(expr)
     else:
         raise TypeError(f"Unable to assemble: {expr}")
 
 
+def _assemble_scalar(expr, scalar, assemble_now, **kwargs):
+    if len(expr.arguments()) != 0:
+        raise ValueError("Can't assemble a 0-form with arguments")
+    if scalar:
+        raise ValueError("Can't assemble 0-form into existing tensor")
+
+    scalar = _make_scalar()
+    _assemble_expr(expr, scalar, diagonal=False, assemble_now=assemble_now, **kwargs)
+    return scalar.data[0]
+
+
+def _assemble_vector(expr, vector, *, diagonal, assemble_now, **kwargs):
+    if diagonal:
+        test, trial = expr.arguments()
+        if test.function_space() != trial.function_space():
+            raise ValueError("Can only assemble diagonal of 2-form if functionspaces match")
+    else:
+        test, = expr.arguments()
+    if vector:
+        if test.function_space() != vector.function_space():
+            raise ValueError("Form's argument does not match provided result tensor")
+        vector.dat.zero()
+    else:
+        vector = _make_vector(test)
+    _assemble_expr(expr, vector, diagonal=diagonal, assemble_now=assemble_now, **kwargs)
+    return vector
```

Comment:
Here, and with the scalar and matrix cases, I'm masking some potentially bad behaviour. Even with `create_assembly_callable()` the callable will always return the tensor (previously it didn't). Since the callable doesn't care about the return value it doesn't cause a problem but it's not ideal.

Personally I'd quite like it if we specified that `None` is returned from `assemble()` if a tensor is provided. That is a significant API change though. We could avoid that by adding a `return_tensor=True` default argument.

Reply:
I think `assemble` should always return the tensor (even if provided). What is the potentially bad behaviour you're masking?

I am happy to futz with the create_assembly_callable API (I would like to just remove it after deprecation).

Reply:
By bad behaviour I mean that the public API has been changed and now the tensor is returned instead of `None`. It's very unlikely to be a problem and deprecating `create_assembly_callable()` will resolve this anyway.

Reply:
Although not named privately, the docstring of `create_assembly_callable` has always indicated that it is not really part of the public API, so I'm kind of happy with this.